edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Service/EdgeHubCertificates.cs

```diff
@@ -33,64 +36,92 @@ public class EdgeHubCertificates
+                    IEnumerable<X509Certificate2> trustBundle = await CertificateHelper.GetTrustBundleFromEdgelet(workloadUri, edgeletApiVersion, Constants.WorkloadApiVersion, moduleId, generationId);
+                    Option<X509Certificate2> manifestTrustBundle = await CertificateHelper.GetManifestTrustBundleFromEdgelet(workloadUri, edgeletApiVersion, Constants.WorkloadApiVersion, moduleId, generationId);
+
+                    result = new EdgeHubCertificates(
+                        certificates.ServerCertificate,
+                        certificates.CertificateChain?.ToList(),
+                        trustBundle?.ToList(),
+                        manifestTrustBundle);
+                }
+                else if (!string.IsNullOrEmpty(edgeHubDevCertPath) &&
+                         !string.IsNullOrEmpty(edgeHubDevPrivateKeyPath) &&
+                         !string.IsNullOrEmpty(edgeHubDevTrustBundlePath))
+                {
+                    // If no connection string was set and we use iotedged workload style certificates for development
+                    (X509Certificate2 ServerCertificate, IEnumerable<X509Certificate2> CertificateChain) certificates;
+
+                    certificates = CertificateHelper.GetServerCertificateAndChainFromFile(edgeHubDevCertPath, edgeHubDevPrivateKeyPath);
+                    IEnumerable<X509Certificate2> trustBundle = CertificateHelper.ParseTrustedBundleFromFile(edgeHubDevTrustBundlePath);
+
+                    result = new EdgeHubCertificates(
+                        certificates.ServerCertificate,
+                        certificates.CertificateChain?.ToList(),
+                        trustBundle?.ToList(),
+                        Option.None<X509Certificate2>());
+                }
+                else if (!string.IsNullOrEmpty(edgeHubDockerCertPFXPath) &&
+                         !string.IsNullOrEmpty(edgeHubDockerCaChainCertPath))
+                {
+                    // If no connection string was set and we use iotedge devdiv style certificates for development
+                    List<X509Certificate2> certificateChain = CertificateHelper.GetServerCACertificatesFromFile(edgeHubDockerCaChainCertPath)?.ToList();
+                    result = new EdgeHubCertificates(new X509Certificate2(edgeHubDockerCertPFXPath), certificateChain, new List<X509Certificate2>(), Option.None<X509Certificate2>());
+                }
+                else
+                {
+                    throw new InvalidOperationException("Edge Hub certificate files incorrectly configured");
+                }
+
+                // On Windows, from time to time the private key cannot be accessed and kestrel fails accepting connections
+                // without the private key. Testing the private key here and if it fails, retry the entire cert import.
+                // The second try usually works.
+                try
+                {
+                    _ = result.ServerCertificate.PrivateKey;
```

Comment:
So the only change is to check if the ServerCertificate has a private key?

Reply:
yes, this is what makes it break the way kestrel would do. As I picture, during connection kestrel does the same to prove the client that is has the private key for its certificate it sent to the client, but fails to obtain it and throws the message we see in the logs.
This call just does it much earlier, so if it would break during connection, we see it now and try to reimport (which magically works somehow)